### ISSUE TITLE
ShellCheck should warn against "return" outside a function

### ISSUE THREAD

**Issue by Screwtapello** [koalaman/shellcheck#2026]
#### For new checks and feature suggestions
- [X] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh
return 42
```

#### Here's what shellcheck currently says:

(no warning)

#### Here's what I wanted or expected to see:

```
Line 2:
return 42
^-- SCXXXX: return cannot reliably be used outside a function

Did you mean: 
exit 42
```

#### Notes

Although (on my machine) dash 0.5.10.2 and zsh 5.8 both accept top-level "return" quite happily, bash 5.0.16 exits with an error:

    path/to/script: line 2: return: can only `return' from a function or sourced script

(the exit status is 1, not 42, so it's an error and not just a warning)

bash behaves the same way when started in POSIX-compatible mode (`--posix`) so on machines where bash is `/bin/sh` this can be a problem even if other `/bin/sh` alternatives are fine with it.

**Comment by ChillerDragon** [comment]
You technically can write a script that returns in global scope and runs fine but you probably shouldn't :D

```
#!/bin/bash

function return() {
    exit "$1"
}

return 1
```

Good issue! Probably useful when refactoring code or similar.

**Comment by stdedos** [comment]
I would like to see the same, but in reverse: Exiting from a function.

Sometimes, sourcing functions which `exit`, they close the GUI container in question, which I've done quite a few times.

I guess it still has legitimate uses (e.g. exiting a recursion?, `(exit 105)` to return a custom status?), so an optional check would be fine by me.

**Comment by ChillerDragon** [comment]
> I would like to see the same, but in reverse: Exiting from a function.
> 
> Sometimes, sourcing functions which `exit`, they close the GUI container in question, which I've done quite a few times.
> 
> I guess it still has legitimate uses (e.g. exiting a recursion?, `(exit 105)` to return a custom status?), so an optional check would be fine by me.

oof not every bash function is sourced and sometimes one wants to abort the whole script in a function

**Comment by stdedos** [comment]
As I said, optionally. I am not using it (_I think_ I am not using), so it would make it a conscious decision of me.

**Comment by ale5000-git** [comment]
I usually use: `return 1 2>&- || exit 1` so it return is sourced otherwise exit.

**Comment by koalaman** [comment]
@SamantazFox Not until there's a strategy for dealing with sourced scripts, where top level `return` is valid

**Comment by stdedos** [comment]
It could always be so that only-sourced scripts are `#!/bin/false`, and they are given their language via a directive. 

Ofc, that's just what _I_ do - it is by no means "an established practice"... 
And as far as I have seen, koalaman avoids setting precedents himself. 😕

**Comment by ale5000-git** [comment]
@stdedos: On some script I made them in a way they works in both ways (sourced and executed).

**Comment by stdedos** [comment]
There's that too 😄

**Comment by ale5000-git** [comment]
But a very clear way would be a shelcheck check directive to set: sourced, executd, both.
